### PR TITLE
v2.0.0 implementation without performance tuning part

### PR DESCRIPTION
**Pull request by mogproject**
# New command-line arguments

Please comment if you come up with better option names and help messages.

```
(cricca) $ python src/main.py
usage: main.py [-h] -a ALGORITHM -g GRAPH_FILENAME -k PARAMETER [-p] [--skip-kernelization] [--profile] [--kernel-v2] [--perf-lp-v2] [-t NUM_THREADS] [-v]
main.py: error: the following arguments are required: -a/--algorithm, -g/--graph_filename, -k/--parameter
(cricca) $ python src/main.py -v
main.py 2.0.0
(cricca) $ python src/main.py -h
usage: main.py [-h] -a ALGORITHM -g GRAPH_FILENAME -k PARAMETER [-p] [--skip-kernelization] [--profile] [--kernel-v2] [--perf-lp-v2] [-t NUM_THREADS] [-v]

Script to run wecp, ipart, or lp on single graph files.

optional arguments:
  -h, --help            show this help message and exit
  -a ALGORITHM, --algorithm ALGORITHM
                        options: wecp, ipart, lp
  -g GRAPH_FILENAME, --graph_filename GRAPH_FILENAME
                        enter .txt edge list data file
  -k PARAMETER, --parameter PARAMETER
                        enter integer parameter value
  -p, --preprocess      enable preprocessing (default: False)
  --skip-kernelization  skip kernelization (default: False)
  --profile             enable profiler (default: False)
  --kernel-v2           use an improved kernelization technique based on the nonuniform Fisher inequality (default: False)
  --perf-lp-v2          use performance tuning techniques for algorithm lp (default: False)
  -t NUM_THREADS, --num-threads NUM_THREADS
                        number of threads used for Gurobi (default: Gurobi's default value)
  -v, --version         show program's version number and exit
```

- `-v`: This version is 2.0.0
- `--profile`: Enables profiler
- `-t` or `--num-threads`: The number of threads used for Gurobi
  - Gurobi sets this value to `min(32, {# of cores in the system})` by default.
  - Of course, more threads don't necessarily mean more performant.

Example:

- Use profiler with the number of threads 80:

```
$ python src/main.py -g data/lv/4_LV_kdistinct5_ktotal17_threshold6_scalefac2_seed0_maxcweight4_n428.txt -a lp -k 4 --profile --num-threads 80
Filename :  data/lv/4_LV_kdistinct5_ktotal17_threshold6_scalefac2_seed0_maxcweight4_n428.txt
kernel time: 4.711190938949585, n_pre=428, n_post=13
Set parameter Username
Academic license - for non-commercial use only - expires 2022-03-12
Set parameter Threads to value 80
Decomposition time:  20.231539011001587

Clique Membership Matrix B:
 [[-1]]
Weight Matrix W:

Decomposition Failed
         23160728 function calls (23145262 primitive calls) in 25.484 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   763877    0.414    0.000    3.957    0.000 <__array_function__ internals>:2(all)
        1    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(amax)
        1    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(amin)
        1    0.000    0.000    0.000    0.000 <__array_function__ internals>:2(argsort)
      429    0.000    0.000    0.002    0.000 <__array_function__ internals>:2(atleast_1d)

(snip)
```

- Enable preprocessing and kernel-v2 (smaller kernel introduced by us)

```
$ python src/main.py -g data/lv/4_LV_kdistinct5_ktotal17_threshold6_scalefac2_seed0_maxcweight4_n428.txt -a lp -k 4 -p --kernel-v2
Filename :  data/lv/4_LV_kdistinct5_ktotal17_threshold6_scalefac2_seed0_maxcweight4_n428.txt

preprocessing, k= 4
preprocess time: 4.324690103530884, n_pre=428, n_post=276
kernel time: 0.587838888168335, infeasible
Decomposition Failed
```

# Bug fixes

- (1) Syntax error when kernelization is skipped.

```
python src/main.py --graph_filename data/lv/4_LV_kdistinct5_ktotal17_threshold6_scalefac2_seed0_maxcweight4_n428.txt --algorithm lp --parameter 5 --kernelize ''
Filename :  data/lv/4_LV_kdistinct5_ktotal17_threshold6_scalefac2_seed0_maxcweight4_n428.txt
Traceback (most recent call last):
  File "src/main.py", line 204, in <module>
    main(args)
  File "src/main.py", line 193, in main
    run(args.algorithm, G, args.parameter, args.preprocess, args.kernelize, args.kernel_v2, args.perf_lp_v2, args.num_threads)
  File "src/main.py", line 135, in run
    A = A.to_numpy()
UnboundLocalError: local variable 'A' referenced before assignment
```

- (2) `argparse` does not support Boolean parameters. They are interpreted as strings.

```
python src/main.py --graph_filename ... --algorithm lp --parameter 5 --kernelize=False
```

This is actually parsed as `args.kernelize = True`

- (3) Runtime error if the kernelization step finds the instance infeasible.
- (4) The program should skip the decomposition step if the kernelization step finds the instance infeasible.

### CODE REVIEW

File: src/algs/kernel.py [hunk 1]
```diff
@@ -35,16 +35,19 @@ def num_blocks(A):
...
     # reduction rule 2
     toremove = []
     b_index=0
+
+    # k: smaller kernel based on the nonuniform Fisher inequality
+    block_size_threshold = k if kernel_v2_enabled else 2 ** k
         
     for block in blocks:
-        if len(block) > 2**k:   
+        if len(block) > block_size_threshold:
```
Comment (mogproject) [inline src/algs/kernel.py:50]: Reduction Rule 1

File: src/main.py [hunk 2]
```diff
@@ -73,26 +76,34 @@ def run_preprocess(G, k_distinct):
...
     end = time.time()   
     time_kernel = end - start
     
-    G.remove_nodes_from(removal_vertices)
-    n_post = G.order()
-    
-    print('kernel time: {}, n_pre={}, n_post={}'.format(time_kernel, n_pre, n_post))
-    return A_kernel
+    # check if the instance is found to be infeasible
+    if removal_vertices != [-1]:
```
Comment (mogproject) [inline src/main.py:95]: bug fix (3)

File: src/main.py [hunk 3]
```diff
@@ -120,19 +131,22 @@ def run_bsd(A, G, k_input, alg_version):
     return B, W
 
 
-def run(alg_version, G, k, to_preprocess, to_kernelize):
+def run(alg_version, G, k, to_preprocess, to_kernelize, kernel_v2_enabled, perf_lp_v2_enabled, num_threads):
     
     if to_preprocess:
         G, k = run_preprocess(G, k)
     
+    A = utils_misc.get_wildcard_adjacency(G)
```
Comment (mogproject) [inline src/main.py:139]: bug fix (1)

File: src/main.py [hunk 4]
```diff
@@ -120,19 +131,22 @@ def run_bsd(A, G, k_input, alg_version):
...
+        if A is None:
+            print("Decomposition Failed")
+            return
+
     A = A.to_numpy()
-    B, W = run_bsd(A, G, k, alg_version)
+    B, W = run_bsd(A, G, k, alg_version, perf_lp_v2_enabled, num_threads)
     
-    print('\nClique Membership Matrix B:\n', B)
+    print('\nClique Membership Matrix B:\n', B)  # FIXME: This may not output the whole results
```
Comment (mogproject) [inline src/main.py:149]: The current version does not output any results. Just (summarized) signatures are shown.

File: src/main.py [hunk 5]
```diff
@@ -163,25 +180,37 @@ def main():
...
-    
-    print("Filename : ", graph_filename)
-    G = read_data(graph_filename)
-    
-    run(algorithm, G, k, to_preprocess, to_kernelize)
-    
+    parser.add_argument('-p', '--preprocess', action='store_true',
+        help="enable preprocessing (default: False)", required=False)
+    parser.add_argument('--skip-kernelization', action='store_true',
+        help="skip kernelization (default: False)", required=False)
```
Comment (mogproject) [inline src/main.py:186]: bug fix (2)

File: src/main.py [hunk 6]
```diff
@@ -163,25 +180,37 @@ def main():
...
-    to_preprocess = args['preprocess']
-    to_kernelize = args['kernelize']
-    
-    print("Filename : ", graph_filename)
-    G = read_data(graph_filename)
-    
-    run(algorithm, G, k, to_preprocess, to_kernelize)
-    
+    parser.add_argument('-p', '--preprocess', action='store_true',
+        help="enable preprocessing (default: False)", required=False)
```
Comment (mogproject) [inline src/main.py:184]: It seems that preprocessing is very effective. Should we run experiments with preprocessing or without, both?

File: src/main.py [hunk 7]
```diff
@@ -120,19 +131,22 @@ def run_bsd(A, G, k_input, alg_version):
...
     
     if to_preprocess:
         G, k = run_preprocess(G, k)
     
+    A = utils_misc.get_wildcard_adjacency(G)
     if to_kernelize:
-        A_wc = utils_misc.get_wildcard_adjacency(G)
-        A = run_kernel(A_wc, G, k)
-        
+        A = run_kernel(A, G, k, kernel_v2_enabled)
```
Comment (sjain12) [inline src/main.py:141]: I'm not super familiar with Python and I am mostly sure this is okay but please confirm that passing A to the function and collecting the return value also in A will not have any adverse effect.

Comment (mogproject) [reply src/main.py:141]: `A_wc` is not used anywhere else, and Python passes references. There should not be any side effects.

**Comment by sjain12** [comment]
Looks good to me!

**Pull request merged**